crates/rnote-ui/src/appwindow/actions.rs

```diff
@@ -724,12 +728,72 @@ impl RnAppWindow {
+                    .coords
+                });
+
+                appwindow.clipboard_paste(last_contextmenu_pos);
+            }),
+        );
+    }
+
+    pub(crate) fn setup_action_accels(&self) {
+        let app = self.app();
+
+        app.set_accels_for_action("win.active-tab-close", &["<Ctrl>w"]);
+        app.set_accels_for_action("win.fullscreen", &["F11"]);
+        app.set_accels_for_action("win.keyboard-shortcuts", &["<Ctrl>question"]);
+        app.set_accels_for_action("win.open-canvasmenu", &["F9"]);
+        app.set_accels_for_action("win.open-appmenu", &["F10"]);
+        app.set_accels_for_action("win.open-doc", &["<Ctrl>o"]);
+        app.set_accels_for_action("win.save-doc", &["<Ctrl>s"]);
+        app.set_accels_for_action("win.save-doc-as", &["<Ctrl><Shift>s"]);
+        app.set_accels_for_action("win.new-tab", &["<Ctrl>t"]);
+        app.set_accels_for_action("win.snap-positions", &["<Ctrl><Shift>p"]);
+        app.set_accels_for_action("win.clear-doc", &["<Ctrl>l"]);
+        app.set_accels_for_action("win.print-doc", &["<Ctrl>p"]);
+        app.set_accels_for_action("win.add-page-to-doc", &["<Ctrl><Shift>a"]);
+        app.set_accels_for_action("win.remove-page-from-doc", &["<Ctrl><Shift>r"]);
+        app.set_accels_for_action("win.zoom-in", &["<Ctrl>plus"]);
+        app.set_accels_for_action("win.zoom-out", &["<Ctrl>minus"]);
+        app.set_accels_for_action("win.import-file", &["<Ctrl>i"]);
+        app.set_accels_for_action("win.undo", &["<Ctrl>z"]);
+        app.set_accels_for_action("win.redo", &["<Ctrl><Shift>z"]);
+        app.set_accels_for_action("win.clipboard-copy", &["<Ctrl>c"]);
+        app.set_accels_for_action("win.clipboard-cut", &["<Ctrl>x"]);
+        app.set_accels_for_action("win.clipboard-paste", &["<Ctrl>v"]);
+        app.set_accels_for_action("win.pen-style::brush", &["<Ctrl>1"]);
+        app.set_accels_for_action("win.pen-style::shaper", &["<Ctrl>2"]);
+        app.set_accels_for_action("win.pen-style::typewriter", &["<Ctrl>3"]);
+        app.set_accels_for_action("win.pen-style::eraser", &["<Ctrl>4"]);
+        app.set_accels_for_action("win.pen-style::selector", &["<Ctrl>5"]);
+        app.set_accels_for_action("win.pen-style::tools", &["<Ctrl>6"]);
+
+        // shortcuts for devel build
+        if config::PROFILE.to_lowercase().as_str() == "devel" {
+            app.set_accels_for_action("win.visual-debug", &["<Ctrl><Shift>v"]);
+        }
+    }
+
+    fn clipboard_paste(&self, target_pos: Option<na::Vector2<f64>>) {
+        let canvas_wrapper = self.active_tab_wrapper();
+        let canvas = canvas_wrapper.canvas();
+        let content_formats = self.clipboard().formats();
+
+        // Order matters here, we want to go from specific -> generic, mostly because `text/plain` is contained in other text based formats
+        if content_formats.contain_mime_type("text/uri-list") {
+            glib::MainContext::default().spawn_local(clone!(@weak self as appwindow => async move {
```

Comment:
I recently replaced these by `glib::spawn_future_local()` which I think is a bit nicer.